CONSOLE-2768: Update default and download deployments to use bindata

florkbr opened the pull request:
Tech debt - migrates our inline manifests for the default console deployment and the download deployment to Go bindata while substituting in necessary variables. This will make it clearer where to modify the base deployments in the console operator (versus looking through Go code). See `bindata/deployments` in the diff.

https://issues.redhat.com/browse/CONSOLE-2768

florkbr commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -64,10 +66,8 @@ type volumeConfig struct {
-	meta := util.SharedMeta()
-	meta.Labels = labels
-	deploymentAnnotations := map[string]string{
+	deployment := resourceread.ReadDeploymentV1OrDie(assets.MustAsset("deployments/console-default-deployment.yaml"))
```
@jhadvig I'm not sure if we will run into the RBAC issue here as these manifests were not managed by the CVO AFAIK. Can you confirm?

jhadvig replied:
we should not run into any RBAC issues here, since its handled by the console-operator itself and it does have proper RBAC for it.

florkbr commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -92,110 +88,50 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
+		{
+			Name:                     "download-server",
+			TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+			Image:                    util.GetImageEnv("DOWNLOADS_IMAGE"),
```
@jhadvig I can move these 2 containers to the bindata, but would need to sub the image info in. Does that seem reasonable to you?

jhadvig replied:
So I would put all the static content in a bindata and the dynamicly generated parts (volumes,...) replace. Check how its done in the [cluster-auth-operator](https://github.com/openshift/cluster-authentication-operator/blob/19971c61b37a47feb2578674f1ae41cc7bee842f/pkg/operator/workload/sync_openshift_oauth_apiserver.go#L134-L152)

florkbr commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -92,110 +88,50 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
+		consoleContainer(operatorConfig, volumeConfig, proxyConfig),
+	}
 	podAnnotations := map[string]string{
 		workloadManagementAnnotation: workloadManagementAnnotationValue,
```
I just realized workloadManagementAnnotationValue is a fixed value in both manifests... I'll move this tomorrow
```
workloadManagementAnnotationValue                    = `{"effect": "PreferredDuringScheduling"}`
```

jhadvig reviewed (commented):
wondering if we will need to rewrite the deployment_test.go since we are not assembling the deployment programatically.

jhadvig commented:
/retest

jhadvig commented on bindata/deployments/console-default-deployment.yaml:
```diff
@@ -0,0 +1,44 @@
+apiVersion: apps/v1
```
lets just call it `console-deployment.yaml`

jhadvig commented on bindata/deployments/console-downloads-deployment.yaml:
```diff
@@ -0,0 +1,40 @@
+apiVersion: apps/v1
```
here as well, lets just call it `downloads-deployment.yaml`

jhadvig reviewed (commented):
@florkbr are you still working on the PR? Or can the WIP be removed?
Found two naming nits.

Also wondering if we still need `TestDefaultDeployment` and `TestDefaultDownloadsDeployment` in the `deployment_test.go`, since the the manifest is now static, except for generating:
- annotations
- replicas
- affinity
- volumes
- containers

Thinking of rather migrating these into helpers(if they are already not) and unit test rather those.

florkbr commented:
@jhadvig I will remove the WIP, but am in the process of making your code review comment changes.

jhadvig commented:
@florkbr let's keep `TestDefaultDeployment` and `TestDefaultDownloadsDeployment` TCs. Sorry for the confusion.

florkbr commented:
@jhadvig this snowballed on me, so I'm adding back the WIP until we sync and confirm this is what you had in mind.

1. I have made your code review suggestions
2. I used the auth operator as an example for string substitution in the yaml. We will now sub out simple things (image URLs)
3. Less simple things that are yaml structured like volume mounts etc are set after we get the deployment struct
4. I modified the UTs minimally to make them pass as an acceptance test. They are pretty useless IMO past seeing them pass once. I would like to refactor them as you stated previously (test the helpers I've added such as `withAffinity` etc)

florkbr commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -64,10 +63,32 @@ type volumeConfig struct {
+	}
+
+	deployment := resourceread.ReadDeploymentV1OrDie(tmpl)
+	withAnnotations(deployment, cm, serviceCAConfigMap, defaultIngressCertConfigMap, trustedCAConfigMap, sec, proxyConfig, infrastructureConfig)
```
@jhadvig any naming preferences here on the helper functions?

jhadvig commented on pkg/console/subresource/deployment/deployment_test.go:
```diff
@@ -9,20 +9,101 @@ import (
+	workloadManagementAnnotationValue = `{"effect": "PreferredDuringScheduling"}`
+)
+
+func consoleContainer(cr *operatorsv1.Console, volConfigList []volumeConfig, proxyConfig *configv1.Proxy) corev1.Container {
```
I dont think we should need this function anymore. Its kinda redundant to the deployment's bindata

florkbr replied:
@jhadvig agreed; removed!

florkbr commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -92,116 +123,50 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
+	if match := re.Find(tmpl); len(match) > 0 {
+		err := fmt.Errorf("invalid template reference %q", string(match))
+		klog.V(4).Infof("%v", err)
+		return nil
```
@jhadvig done!

florkbr commented:
@jhadvig please re-review. I've addressed your comments as well as added unit tests for each of the new helper functions.

florkbr commented on pkg/console/subresource/deployment/deployment_test.go:
```diff
@@ -366,13 +391,494 @@ func TestDefaultDeployment(t *testing.T) {
+			args: args{
+				deployment: &appsv1.Deployment{
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{},
```
@jhadvig one question I had: should these helper functions handle the case where `Deployment.ObjectMeta.Annotations` has not been initialized to an empty string map? Or are we guaranteed these will always be initialized?

jhadvig replied:
Since console's deployments are relying on other resources(infraConfig, ingress, various CMs, ....) and generating its deployments from them, we expect them to exist and be available. Otherwise operator gets degraded.

jhadvig commented:
/retest

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -77,13 +82,23 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
-	// Set any annotations as needed so that `ApplyDeployment` rolls out a
-	// new version when they change.
-	meta.Annotations = deploymentAnnotations
+func withReplicas(deployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
```
So I though that this will either replace the `Replicas()` function, since it doesnt make sense to have a function for determining the proper value(replica count in this set) and its wrapper thats being a setter. We should combine it into a single function. Also this way we would just keep enlarge our unit test cases without any real added value.

florkbr replied:
@jhadvig done!

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -77,13 +82,23 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
+	deployment.Spec.Replicas = &replicas
+}
+
+func withAffinity(deployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
```
same as above

florkbr replied:
@jhadvig done!

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -92,116 +107,41 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
 	return downloadsDeployment
 }
 
+func withDownloadsReplicas(downloadsDeployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
```
do we need a separate function for console (`withReplicas()`) and for downloads (`withDownloadsReplicas()`) ?

florkbr replied:
@jhadvig done!

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -92,116 +107,41 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
+	downloadsDeployment.Spec.Template.Spec.Affinity = downloadsPodAffinity(infrastructureConfig)
+}
+
+func withDownloadsContainers(downloadsDeployment *appsv1.Deployment) {
```
```suggestion
func withDownloadsContainerImage(downloadsDeployment *appsv1.Deployment) {
```

florkbr replied:
@jhadvig done!

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -92,116 +107,41 @@ func DefaultDeployment(operatorConfig *operatorv1.Console, cm *corev1.ConfigMap,
+	downloadsDeployment.Spec.Replicas = &replicas
+}
+
+func withDownloadsAffinity(downloadsDeployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
```
same as above affinity function comment

florkbr replied:
@jhadvig done!

jhadvig reviewed (commented):
@florkbr we are getting closer. Added couple of comments. The point is not to enlarge our codebase/testCases but rather refactor them. PTAL

jhadvig commented:
@florkbr PR needs rebase.

florkbr commented:
@jhadvig third time is the charm hopefully :) Could you please have another look?

florkbr commented:
/retest

florkbr commented:
/retest

jhadvig commented:
/test e2e-aws-operator

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -367,19 +193,85 @@ func consoleVolumes(vc []volumeConfig) []corev1.Volume {
-			Name:      item.name,
-			ReadOnly:  item.readOnly,
-			MountPath: item.path,
+func withContainerImage(deployment *appsv1.Deployment, operatorConfig *operatorv1.Console, proxyConfig *configv1.Proxy) {
```
Same as above, since it's not deployment agnostic.
```suggestion
func withConsoleContainerImage(deployment *appsv1.Deployment, operatorConfig *operatorv1.Console, proxyConfig *configv1.Proxy) {
```

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -367,19 +193,85 @@ func consoleVolumes(vc []volumeConfig) []corev1.Volume {
+}
+
+func withStrategy(deployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
+	var rollingUpdateParams *appsv1.RollingUpdateDeployment
```
nit: no need to initialise the var:
```go
	rollingUpdateParams := &appsv1.RollingUpdateDeployment{}
	if infrastructureConfig.Status.InfrastructureTopology != configv1.SingleReplicaTopologyMode {
		rollingUpdateParams = &appsv1.RollingUpdateDeployment{
			MaxSurge: &intstr.IntOrString{
				IntVal: int32(3),
			},
			MaxUnavailable: &intstr.IntOrString{
				IntVal: int32(1),
			},
		}
	}
       deployment.Spec.Strategy.RollingUpdate = rollingUpdateParams
```

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -367,19 +193,85 @@ func consoleVolumes(vc []volumeConfig) []corev1.Volume {
+	deployment.Spec.Strategy.RollingUpdate = rollingUpdateParams
+}
+
+func withNodeSelector(deployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
```
```suggestion
func withConsoleNodeSelector(deployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
```

jhadvig commented on pkg/console/subresource/deployment/deployment_test.go:
```diff
@@ -142,7 +151,19 @@ func TestDefaultDeployment(t *testing.T) {
+						},
+					},
+				},
+				TopologyKey: "topology.kubernetes.io/zone",
```
We got rid of this rule

jhadvig commented on pkg/console/subresource/deployment/deployment_test.go:
```diff
@@ -165,8 +186,14 @@ func TestDefaultDeployment(t *testing.T) {
+	consoleDeploymentContainer := consoleDeploymentTemplate.Spec.Template.Spec.Containers[0]
+	consoleDeploymentVolumes := consoleDeploymentTemplate.Spec.Template.Spec.Volumes
+	withVolumes(consoleDeploymentTemplate, trustedCAConfigMapSet, false)
+	consoleDeploymentContainerTrusted := consoleDeploymentTemplate.Spec.Template.Spec.Containers[0]
```
Why `...Trusted` ?

jhadvig commented on pkg/console/subresource/deployment/deployment_test.go:
```diff
@@ -916,95 +1077,562 @@ func Test_consoleVolumes(t *testing.T) {
+			},
+		},
+	}
+	downloadsDeploymentPodSpecHighAvail := downloadsDeploymentPodSpecSingleReplica
```
not new but its a good practise 
```suggestion
	downloadsDeploymentPodSpecHighAvail := downloadsDeploymentPodSpecSingleReplica.DeepCopy()
```

jhadvig commented on pkg/console/subresource/deployment/deployment_test.go:
```diff
@@ -916,95 +1077,562 @@ func Test_consoleVolumes(t *testing.T) {
+						},
+					},
+				},
+				TopologyKey: "topology.kubernetes.io/zone",
```
Lets remove this rule here as well

jhadvig reviewed (commented):
@florkbr I've added some additional comments + nits. PTAL

florkbr commented:
/retest-required

jhadvig commented on pkg/console/subresource/deployment/deployment.go:
```diff
@@ -367,19 +202,58 @@ func consoleVolumes(vc []volumeConfig) []corev1.Volume {
+	deployment.Spec.Template.Spec.Containers[0].Image = util.GetImageEnv("CONSOLE_IMAGE")
+}
+
+func withConsoleNodeSelector(deployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
```
nit: can be just

```suggestion
func withConsoleNodeSelector(deployment *appsv1.Deployment, infrastructureConfig *configv1.Infrastructure) {
	// If running with an externalized control plane, remove the master node selector
	if infrastructureConfig.Status.ControlPlaneTopology == configv1.ExternalTopologyMode {
		deployment.Spec.Template.Spec.NodeSelector = map[string]string{}
	}
}
```

jhadvig reviewed (approved):
@florkbr thanks for patience! Found one small nit, otherwise the PR looks good 👍 
/approve
/lgtm

florkbr commented:
/retest-required

florkbr commented:
/retest

jhadvig commented:
/lgtm

jhadvig commented:
/hold for approvals
/assign @yapei @sferich888 @ahardin-rh

RickJWagner commented:
/lgtm
/label px-approved

opayne1 commented:
/label docs-approved

spadgett commented:
/hold cancel
tide ensures the approved labels are present

yanpzhan commented:
/label qe-approved
